### PR TITLE
Add test for ontology describe view

### PR DESCRIPTION
**Pull request by nfusionz**

### CODE REVIEW

File: tests/api/test_nanopubs.py [hunk 1]
```diff
@@ -100,3 +99,45 @@ def test_attribute_view(self):
         self.assertEquals(json_content['label'], "Jane Doe")
         self.assertEquals(len(json_content['type']), 1)
         self.assertEquals(json_content['type'][0]['label'], 'Person')
+
+    def test_ontology_describe_view(self):
```
Comment (gordom6) [inline tests/api/test_nanopubs.py]: This test doesn't really belong here, but I don't have a better place for it at the moment.

File: tests/api/test_nanopubs.py [hunk 2]
```diff
@@ -100,3 +99,45 @@ def test_attribute_view(self):
...
+        self.assertIn("@id", data[0], "'Describe' view missing @id key")
+
+        graph = data[0]["@graph"]
+        self.assertTrue(graph, "'Describe' view returned no subjects")
+
+        for subject in graph:
+            if subject["@id"] == "http://example.com/":
+                self.assertEquals(len(subject.keys()), 2,
+                                  "Subject in 'describe' view has unexpected number of predicates")
+            elif subject["@id"] == "http://example.com/janedoe":
```
Comment (gordom6) [inline tests/api/test_nanopubs.py]: Cover other subjects?

File: tests/api/test_nanopubs.py [hunk 3]
```diff
@@ -100,3 +99,45 @@ def test_attribute_view(self):
...
+
+        self.login(*self.create_user("user@example.com", "password"))
+        response = self.client.post("/pub",
+                                    data=ontology,
+                                    content_type="text/turtle",
+                                    follow_redirects=True)
+        self.assertStatus(response, 201,
+                          "Did not receive '201 CREATED' when publishing ontology nanopub")
+
+        content = self.client.get("/about",
```
Comment (gordom6) [inline tests/api/test_nanopubs.py]: Lines 120-125 should be factored out into a reusable method on ApiTestCase. Pass in the view and the (optional) template name.

File: tests/api/test_nanopubs.py [hunk 4]
```diff
@@ -100,3 +99,45 @@ def test_attribute_view(self):
...
+    def test_ontology_describe_view(self):
+        ontology = """
+        <http://example.com/> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Ontology> .
+        <http://example.com/janedoe> <http://schema.org/jobTitle> "Professor";
+        <http://schema.org/name> "Jane Doe" ;
+        <http://schema.org/telephone> "(425) 123-4567" ;
+        <http://schema.org/url> <http://www.janedoe.com> ;
+        <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> . """
+
+        self.login(*self.create_user("user@example.com", "password"))
```
Comment (gordom6) [inline tests/api/test_nanopubs.py]: 112-117 should also be factored out; look for reuse

File: tests/api/test_nanopubs.py [hunk 5]
```diff
@@ -1,27 +1,22 @@
...
-    
-    def test_create(self):
-        self.login(*self.create_user("user@example.com","password"))
-
-        response = self.client.post("/pub", data=self.turtle, content_type="text/turtle",follow_redirects=True)
-        
-        self.assertEquals(response.status,'201 CREATED')
 
+    def test_create(self):
+        response = self.publish_nanopub(self.turtle, "text/turtle")
```
Comment (gordom6) [inline tests/api/test_nanopubs.py]: Use a keyword only declaration

def publish_nanopub(self, *, content_type: str, data: str):
...

Then you have to do data=self.turtle, content_type="text/turtle".

Positional arguments should only be used for single-argument callables (functions, methods, etc.) or very well-known ones in the standard library.

File: tests/api/test_nanopubs.py [hunk 6]
```diff
@@ -31,71 +26,54 @@ def test_create(self):
...
         self.assertEquals(len(g), 17)
         self.assertEquals(g.value(URIRef('http://example.com/janedoe'), RDF.type), URIRef('http://schema.org/Person'))
 
     def test_delete_admin(self):
-        self.login(*self.create_user("user@example.com","password"))
-
-        response = self.client.post("/pub", data=self.turtle, content_type="text/turtle",follow_redirects=True)
-        self.assertEquals(response.status,'201 CREATED')
+        response = self.publish_nanopub(self.turtle, "text/turtle")
         self.assertTrue('Location' in response.headers)
```
Comment (gordom6) [inline tests/api/test_nanopubs.py]: Repeated code again?

File: whyis/test/api_test_case.py [hunk 7]
```diff
@@ -2,4 +2,29 @@
 
 
 class ApiTestCase(TestCase):
-    pass
+
+    def publish_nanopub(self, data, content_type, expected_response=201, login=True):
+        if login:
```
Comment (gordom6) [inline whyis/test/api_test_case.py]: login should still be a separate operation. Functions should not change their behavior substantially based on boolean flags.

File: whyis/test/api_test_case.py [hunk 8]
```diff
@@ -2,4 +2,29 @@
...
+
+        response = self.client.post("/pub", data=data,
+                                    content_type=content_type,
+                                    follow_redirects=True)
+        self.assertStatus(response, expected_response,
+                          "Expected {}, got {} as reponse".format(expected_response, response))
+
+        return response
+        
+    def get_view(self, uri, view=None, expected_mime=None, expected_template=None):
```
Comment (gordom6) [inline whyis/test/api_test_case.py]: Another keyword-only, self, *, ...

File: whyis/test/api_test_case.py [hunk 9]
```diff
@@ -2,4 +2,29 @@
...
+        content = self.client.get("/about",
+                                  query_string={ a: b for a, b in zip(["uri", "view"], [uri, view]) if b is not None },
+                                  follow_redirects=True)
+        
+        if expected_template is not None:
+            self.assertTemplateUsed(expected_template)
+
+        if expected_mime is not None:
```
Comment (gordom6) [inline whyis/test/api_test_case.py]: mime_type

File: tests/api/test_nanopubs.py [hunk 10]
```diff
@@ -31,71 +26,54 @@ def test_create(self):
...
         self.assertEquals(content.mimetype, "application/json")
-        g.parse(data=str(content.data,'utf8'), format="json-ld")
-        
+        g.parse(data=str(content.data, 'utf8'), format="json-ld")
+
         self.assertEquals(len(g), 17)
         self.assertEquals(g.value(URIRef('http://example.com/janedoe'), RDF.type), URIRef('http://schema.org/Person'))
 
     def test_delete_admin(self):
-        self.login(*self.create_user("user@example.com","password"))
```
Comment (gordom6) [inline tests/api/test_nanopubs.py]: A good pattern here is a sort of ghetto partial evaluation: have a public method that takes a bunch of parameters and then a private method that fills some of them in.
def login on the parent class, unchanged
def __login(self): return self.login(*self.create_user("user@example.com", "password"))

File: whyis/test/api_test_case.py (hunk 9 above)
Comment (gordom6) [inline whyis/test/api_test_case.py]: The MIME type should probably be required, or?

Comment (jimmccusker) [reply whyis/test/api_test_case.py]: By default, it serves linked data on the entity. If HTML is Accept-able, then the view itself is rendered.

**Pull request merged**